File: web/apps/app/lib/core/chat/chat-capability-outbound-decision.test.ts
import { describe, expect, it } from 'vitest';
import { resolveChatCapabilityOutboundDecision } from './chat-capability-outbound-decision';

describe('ChatCapabilityOutboundDecision', () => {
  const available = { state: 'auto_available' as const, availableIntents: ['off', 'low', 'balanced', 'deep', 'max'] };

  it('reflects the effective outbound capability selection', () => {
    expect(resolveChatCapabilityOutboundDecision({
      webRequested: true,
      webControl: available,
      webDormant: false,
      reasoningModeRequested: 'deep',
      reasoningControl: available,
      reasoningDormant: false,
    })).toMatchObject({
      webSearchEnabled: true,
      reasoningMode: 'deep',
      reasoningIntent: 'deep',
      hasActiveCapabilitySelection: true,
    });
  });

  it('does not highlight stored values suppressed by dormant or allowlist gates', () => {
    expect(resolveChatCapabilityOutboundDecision({
      webRequested: true,
      webControl: available,
      webDormant: true,
      reasoningModeRequested: 'fast',
      reasoningControl: { ...available, availableIntents: ['deep'] },
      reasoningDormant: false,
    })).toMatchObject({
      webSearchEnabled: false,
      reasoningMode: 'automatic',
      hasActiveCapabilitySelection: false,
    });
  });
});

File: android/app/src/test/java/ai/oriveo/community/feature/chat/ChatCapabilityOutboundDecisionTest.kt
package ai.oriveo.community.feature.chat

import ai.oriveo.community.core.model.CapabilityPreferenceValues
import ai.oriveo.community.core.model.CapabilityWebPreference
import ai.oriveo.community.core.model.ReasoningMode
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class ChatCapabilityOutboundDecisionTest {
    private val controls = ChatModelCapabilityResolver.ModelControls(
        webAvailable = true,
        webForceAvailable = true,
        reasoningIntents = listOf("off", "low", "balanced", "deep", "max"),
        reasoningState = "auto_available",
    )

    @Test
    fun `active selections mirror the effective outbound request`() {
        val decision = ChatCapabilityOutboundDecision.resolve(
            CapabilityPreferenceValues(CapabilityWebPreference.Automatic, "deep"),
            controls,
        )

        assertTrue(decision.webSearchEnabled)
        assertEquals(ReasoningMode.Deep, decision.reasoningMode)
        assertEquals("deep", decision.reasoningIntent)
        assertTrue(decision.hasActiveCapabilitySelection)
    }

    @Test
    fun `dormant owners and unavailable intents do not highlight the chip`() {
        val dormant = ChatCapabilityOutboundDecision.resolve(
            CapabilityPreferenceValues(CapabilityWebPreference.Automatic, "deep"),
            controls,
            dormantOwners = setOf("web", "reasoning"),
        )
        val unavailableIntent = ChatCapabilityOutboundDecision.resolve(
            CapabilityPreferenceValues(CapabilityWebPreference.Off, "low"),
            controls.copy(reasoningIntents = listOf("deep")),
        )

        assertFalse(dormant.hasActiveCapabilitySelection)
        assertEquals(ReasoningMode.Automatic, dormant.reasoningMode)
        assertFalse(unavailableIntent.hasReasoningSelection)
    }
}
